### PR TITLE
fix queue hard constraints don't work

### PR DESCRIPTION
**Pull request by yahaa**
Try to fix https://github.com/volcano-sh/volcano/issues/1463

### CODE REVIEW

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 1]
```diff
@@ -241,7 +245,7 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
 		queue := obj.(*api.QueueInfo)
 		attr := pp.queueOpts[queue.UID]
 
-		overused := !attr.allocated.LessEqual(attr.deserved)
```
Comment (yahaa) [inline pkg/scheduler/plugins/proportion/proportion.go:244]: When allocated <= deserved i think we should set queue as overused.

Comment (shinytang6) [reply pkg/scheduler/plugins/proportion/proportion.go:244]: There are still serious logic problems  in func `Less`, I think this change will introduce more problems

ref: https://github.com/volcano-sh/volcano/issues/1036

File: pkg/scheduler/actions/allocate/allocate.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ package allocate
 import (
 	"k8s.io/klog"
 	"volcano.sh/apis/pkg/apis/scheduling"
```
[the remark this answers is not in the document]
Comment (yahaa) [reply pkg/scheduler/actions/allocate/allocate.go:22]: ok, sorted it.

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 1 above)
Comment (yahaa) [reply pkg/scheduler/plugins/proportion/proportion.go:244]: > There are still serious logic problems in func `Less`, I think this change will introduce more problems
> 
> ref: #1036

I try to reimplement `Less` function, and i have add some boundary test cases, please review again.

Comment (Thor-wl) [reply pkg/scheduler/plugins/proportion/proportion.go:244]: > There are still serious logic problems in func `Less`, I think this change will introduce more problems
> 
> ref: #1036

Maybe we should give a clear definition about Less and LessEqual for resources. @shinytang6

File: pkg/scheduler/api/resource_info.go [hunk 3]
```diff
@@ -238,35 +238,76 @@ func (r *Resource) Less(rr *Resource) bool {
 		return l < r
 	}
 
-	if !lessFunc(r.MilliCPU, rr.MilliCPU) {
-		return false
+	left := make(map[v1.ResourceName]float64)
```
[the remark this answers is not in the document]
Comment (yahaa) [reply pkg/scheduler/api/resource_info.go:241]: Yes, `only r.resources at all dimensions is less than rr can this funtion return true.` 

I think we need to make a clear definition about Less functions.

If it's definitions is as what i mentioned above, this case `r = {cpu: 1, memory:2, gpu: 3}, rr = {cpu: 1, memory:3, gpu: 3}` should return false.

File: pkg/scheduler/api/resource_info_test.go [hunk 4]
```diff
@@ -403,6 +422,55 @@ func TestLess(t *testing.T) {
...
+			},
+			resource2: &Resource{
+				MilliCPU:        8000,
+				Memory:          8000,
+				ScalarResources: map[v1.ResourceName]float64{"scalar.test/scalar1": 0, "hugepages-test": 0},
+			},
+			expected: true,
+		},
+		{
+			resource1: &Resource{
```
[the remark this answers is not in the document]
Comment (yahaa) [reply pkg/scheduler/api/resource_info_test.go:451]: It depend on the definitions of `Less`

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 1 above)
Comment (Thor-wl) [reply pkg/scheduler/plugins/proportion/proportion.go:244]: Let me add comments about the definition of Less/LessEqual for all resource comparation functions.
/assign @Thor-wl

File: pkg/scheduler/api/resource_info_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (shinytang6) [reply pkg/scheduler/api/resource_info_test.go:451]: Yes, it should return true IMO.

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 1 above)
Comment (shinytang6) [reply pkg/scheduler/plugins/proportion/proportion.go:244]: The changes in `Less` still can't resolve the problem l list above. In commit https://github.com/volcano-sh/volcano/pull/1350/commits/bd78d79abb1fbdb086bea0760bf4ea6930aabc8a, l have rewrite the `Less` function using math.big, maybe we can merge that one first?

Comment (Thor-wl) [reply pkg/scheduler/plugins/proportion/proportion.go:244]: > The changes in `Less` still can't resolve the problem l list above. In commit [bd78d79](https://github.com/volcano-sh/volcano/commit/bd78d79abb1fbdb086bea0760bf4ea6930aabc8a), l have rewrite the `Less` function using math.big, maybe we can merge that one first?

I'm ok about that.

**Comment by Thor-wl** [comment]
I'll merge this PR after accurate definitions for all resource comparation functions and big.rat is adopted in https://github.com/volcano-sh/volcano/pull/1350

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 1 above)
Comment (Thor-wl) [reply pkg/scheduler/plugins/proportion/proportion.go:244]: > Let me add comments about the definition of Less/LessEqual for all resource comparation functions in volcano.sh/volcano/pkg/scheduler/api/resource_info.go.
> /assign @Thor-wl

ref: https://github.com/volcano-sh/volcano/pull/1490

**Comment by Thor-wl** [comment]
https://github.com/volcano-sh/volcano/pull/1350 is merged

File: pkg/scheduler/api/resource_info.go [hunk 5]
```diff
@@ -368,6 +368,34 @@ func (r *Resource) LessEqualStrict(rr *Resource) bool {
 	return true
 }
 
+// BiggerEqual returns true only on the following conditions:
+// 1. All dimensions resources in r are bigger than that of rr
+// 2. Part dimensions are equal while the others are bigger in r
+// 3. All dimensions resources in r are equal with that of rr
+// Otherwise returns false.
+// Note: Any dimension of resource, which is not listed in resource object, is regarded as zero.
+func (r *Resource) BiggerEqual(rr *Resource) bool {
```
Comment (Thor-wl) [inline pkg/scheduler/api/resource_info.go:377]: just a tip, GreaterEqual may be better

Comment (yahaa) [reply pkg/scheduler/api/resource_info.go]: Done

File: pkg/scheduler/api/resource_info.go [hunk 6]
```diff
@@ -368,6 +368,34 @@ func (r *Resource) LessEqualStrict(rr *Resource) bool {
...
+	if !biggerEqualFn(r.MilliCPU, rr.MilliCPU) {
+		return false
+	}
+	if !biggerEqualFn(r.Memory, rr.Memory) {
+		return false
+	}
+
+	for rName, rQuant := range r.ScalarResources {
+		_, ok := rr.ScalarResources[rName]
+		if !ok || !biggerEqualFn(rQuant, rr.ScalarResources[rName]) {
```
Comment (shinytang6) [inline pkg/scheduler/api/resource_info.go:391]: why `!ok` it returns false?

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 7]
```diff
@@ -162,18 +163,21 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
 			attr.deserved.Add(remaining.Clone().Multi(float64(attr.weight) / float64(totalWeight)))
 
 			if attr.capability != nil && !attr.deserved.LessEqualStrict(attr.capability) {
-				attr.deserved = helpers.Min(attr.deserved, attr.capability)
-				attr.deserved = helpers.Min(attr.deserved, attr.request)
```
Comment (shinytang6) [inline pkg/scheduler/plugins/proportion/proportion.go:166]: does it conflict with https://github.com/volcano-sh/volcano/pull/1550
which one should be reviewed? /cc @Thor-wl

**Review by shinytang6 — CHANGES_REQUESTED** [review]

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 8]
```diff
@@ -162,18 +163,21 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
...
 				klog.V(4).Infof("queue <%s> is meet cause of the capability", attr.name)
-			} else if attr.request.LessEqualStrict(attr.deserved) {
-				attr.deserved = helpers.Min(attr.deserved, attr.request)
-				meet[attr.queueID] = struct{}{}
-				klog.V(4).Infof("queue <%s> is meet", attr.name)
-			} else {
-				attr.deserved.MinDimensionResource(attr.request)
-				klog.V(4).Infof("Format queue <%s> deserved resource to <%v>", attr.name, attr.deserved)
 			}
+			//else if attr.request.LessEqualStrict(attr.deserved) {
```
Comment (shinytang6) [inline pkg/scheduler/plugins/proportion/proportion.go:172]: does these changes related with the issue, can you explain more?

**Review by shinytang6 — CHANGES_REQUESTED** [review]

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 7 above)
Comment (Thor-wl) [reply pkg/scheduler/plugins/proportion/proportion.go:166]: > does it conflict with #1550
> which one should be reviewed? /cc @Thor-wl

PR1550 will be reconstructed after https://github.com/volcano-sh/volcano/issues/1525 is reviewed and settled. But these modifications should be given more explaination.

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 8 above)
Comment (yahaa) [reply pkg/scheduler/plugins/proportion/proportion.go]: This logic is a bit confuse, but it is not related to this issue. I have keep it as the original.

File: pkg/scheduler/api/resource_info.go (hunk 6 above)
Comment (yahaa) [reply pkg/scheduler/api/resource_info.go]: Fix, and add some ut

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 8 above)
Comment (shinytang6) [reply pkg/scheduler/plugins/proportion/proportion.go]: Actually I also want to know why the judgment `attr.request.LessEqualStrict(attr.deserved)` is needed.

Comment (yahaa) [reply pkg/scheduler/plugins/proportion/proportion.go]: That is what i want to know too 😂 ,  i think we can fix it in the next PR.

Comment (shinytang6) [reply pkg/scheduler/plugins/proportion/proportion.go]: l think so. Any thoughts why this judgment is needed? /cc @Thor-wl @k82cn

Comment (yahaa) [reply pkg/scheduler/plugins/proportion/proportion.go]: /cc @Thor-wl

Comment (Thor-wl) [reply pkg/scheduler/plugins/proportion/proportion.go]: > /cc @Thor-wl

I'm sorry for not have a better understanding.

**Pull request closed without merging**